src/torcontrol.cpp

```diff
@@ -326,6 +326,53 @@ TorController::~TorController()
     }
 }
 
+void TorController::get_socks_cb(TorControlConnection& _conn, const TorControlReply& reply)
+{
+    // NOTE: We can only get here if -onion is unset
+    std::string socks_location;
+    if (reply.code == 250) {
+        for (const auto& line : reply.lines) {
+            if (0 == line.compare(0, 20, "net/listeners/socks=")) {
+                const std::string port_list_str = line.substr(20);
+                std::vector<std::string> port_list;
+                boost::split(port_list, port_list_str, boost::is_any_of(" "));
+                for (auto& portstr : port_list) {
+                    if (portstr.empty()) continue;
+                    if ((portstr[0] == '"' || portstr[0] == '\'') && (*portstr.rbegin() == portstr[0])) {
+                        portstr = portstr.substr(1, portstr.size() - 2);
+                        if (portstr.empty()) continue;
+                    }
+                    socks_location = portstr;
+                    if (0 == portstr.compare(0, 10, "127.0.0.1:")) {
+                        // Prefer localhost - ignore other ports
+                        break;
+                    }
+                }
+            }
+        }
+        if (!socks_location.empty()) {
+            LogPrint(BCLog::TOR, "tor: Get SOCKS port command yielded %s\n", socks_location);
+        } else {
+            LogPrintf("tor: Get SOCKS port command returned nothing\n");
+        }
+    } else if (reply.code == 510) {  // 510 Unrecognized command
+        LogPrintf("tor: Get SOCKS port command failed with unrecognized command (You probably should upgrade Tor)\n");
+    } else {
+        LogPrintf("tor: Get SOCKS port command failed; error code %d\n", reply.code);
+    }
+
+    if (socks_location.empty()) {
+        // Fallback to old behaviour
+        socks_location = "127.0.0.1";
+    }
+
+    CService resolved(LookupNumeric(socks_location, 9050));
```

Comment:
Port 9150 for Windows: https://github.com/bitcoin/bitcoin/pull/22651#discussion_r684460683

Reply:
Is it really Windows default? Not confusing with Tor Browser's default, which is 9150?

Reply:
> Not confusing with Tor Browser's default, which is 9150?

Yes this is what I normally follow:

Windows: Running Tor browser and using it as proxy
Linux: Running Tor after installing with command: `sudo apt install tor`

Running Tor as service on Windows involves some workaround which I am not sure many do or maybe I am not aware of other easier ways to do it: https://superuser.com/questions/1631178/how-to-configure-tor-as-service-on-windows

Reply:
9050 here serves two purposes:

1. Fallback to old behaviour. This was always 9050.
2. Default port if Tor's API returns just an IP. Is there any reason to think it would ever be anything other than 9050?

Reply:
Here is my torcfg settings

```
## Tor opens a SOCKS proxy on port 9050 by default -- even if you don't
## configure one below. Set "SOCKSPort 0" if you plan to run Tor only
## as a relay, and not make any local application connections yourself.
SOCKSPort 9050 # Default: Bind to localhost:9050 for local connections.
SOCKSPort 192.168.1.2:9050 # Bind to this address:port too.
```

Note the following things:
1. 9050 is the default value. There is no reason to assume that it will be 9050. It could be anything. The port number is included in the reply. It does not feel right to have it in the reply and assume that it will be 9050 and hardcode that value instead of parsing it.
2. If only the port is provided, it is binded implicitly to 127.0.0.1. I need to test this patch to see the actual output but if what it printed by nyx is what is returned as-is by torcontrol, the config above would return:

`9050, 192.168.1.2:9050`

Your function wants to favor 127.0.0.1 but I think that it could miss the standalone port value scenario...

Reply:
I have few more thoughts to share to possibly improve further the code of this function. Here is my setup:

I have 1 tor proxy configured on a single machine in my LAN. It is on 192.168.1.2.
Bitcoind is running on a different machine: 192.168.1.179.

My first attempt to make torcontrol available for bitcoind was to bind the control listening socket to 192.168.1.2:9051.
It works except that tor is complaining big time that this is opening a security risk as the control protocol exchange is in clear text over the network.

To fix that, I did create a SSH tunnel between the 2 machines 127.0.0.1:9051.

What that means is that when I start bitcoind from 192.168.1.179 machine with the cmdline: -proxy 192.168.1.2:9050 -torcontrol 127.0.0.1:9051

When bitcoind connects with torcontrol, despite it is a localhost address, it does actually connect to a remote server.

What all this means for this commit?

Well, it favors the tor proxy address 127.0.0.1. With my setup, it will fail.

The first solution to address this issue that came to my mind is to test that 127.0.0.1:9050 is listening but this is clumsy, there is no simple and portable way to check that beside actually connect a socket to the address...

Then, I did ask myself the following question? What are exactly the benefits of connecting to the localhost proxy address if present?

I am not sure if there is any.

OTOH, if you connect to the first public IP address listed in the reply, it will work with the most common case where the proxy is on the same machine and it will also work on more funky setups like mine...

How do you feel toward this suggestion? Does that makes sense to you?

Reply:
I tested with
```
SOCKSPort 10000
SOCKSPort 192.168.0.1:20000
SOCKSPort 0.0.0.0:30000
```

The reply from the Tor control daemon is:
```
net/listeners/socks="127.0.0.1:10000" "192.168.0.1:20000" "0.0.0.0:30000"
```
so it does not return just port. The spec is here: https://github.com/torproject/torspec/blob/845a4d7213e8e28bb039d6925437b5b1c0d607e7/control-spec.txt#L912.

Maybe it makes sense to choose the non-`127.0.0.1` address if there are a few:
1. If the Tor daemon is on the same machine as bitcoind, then either one will work and communication will be local anyway
2. If the Tor daemon is on another machine, only the non-`127.0.0.1` address will work.

But anyway - we can't expect the automatic configuration to detect and work in all cases. For this we have the manual override. @lano1106, in your case you can set `-onion=192.168.1.2:9050` which will work with `master` and with this PR.

Reply:
Vasil, you have been faster than me... I just tested too this morning:
`2021-09-02T13:51:22Z tor: reply line: net/listeners/socks="127.0.0.1:9050" "192.168.1.2:9050"`

So overall this pull request works well. I just think non localhost address should be prefered over 127.0.0.1. This would work with my setup and I don't any benefits of picking 127.0.0.1.

Reply:
>So overall this pull request works well. I just think non localhost address should be prefered over 127.0.0.1. This would work with my setup and I don't see any extra benefits of picking 127.0.0.1 in any scenario.

The non-localhost address may change or cease to be available?

>because -onion switch is not the only way to configure the onion proxy.

But it's the only way that should override `-torcontrol` IMO

Reply:
@luke-jr if the non-localhost address may change or cease to be available is a serious and valid concern for not modifying the code behavior then so be it.

I have said everything that I have to say on the topic and I'll conclude by saying that without being aware of the initial problem that motivated this pull request, with the reasonable assumption that the vast majority of tor proxy setups will list the localhost address, this pull request, if left as-is, will just be a shinier, more complex method to essentially reach the exact same result than what the current code is doing...